Implement typed unit IDs

devreal opened the pull request:
Implement typed unit IDs to differentiate between global (rare) and team-local unit IDs (common). Spotted some wrong usages in the process of converting most places where `dart_unit_t` was used. 

Addresses #143

devreal committed:
Implement a generic unitid struct and operations on it; provide specialization for global and local unit ID

devreal committed:
Implement a generic unitid struct and operations on it; provide specialization for global and local unit ID

devreal committed:
Start integration of typed unit IDs with GlobMem

devreal committed:
Start integration of typed unit IDs with GlobMem

devreal committed:
Adapt HDF5 test to typed unit IDs

devreal committed:
Adapt HDF5 test to typed unit IDs

devreal committed:
Add static function dash::Team::GlobalUnitID and implement dash::myid() in terms of that

devreal committed:
Add static function dash::Team::GlobalUnitID and implement dash::myid() in terms of that

devreal committed:
Make unit_id.id private and add friend functions; Remove division operators

devreal committed:
Make unit_id.id private and add friend functions; Remove division operators

devreal committed:
Fix type in dash::find_if and dash::equal

devreal committed:
Fix type in dash::find_if and dash::equal

devreal committed:
Convert GlobDynamicMem to typed unit IDs

devreal committed:
Convert GlobDynamicMem to typed unit IDs

devreal committed:
Add DASH_UNDEFINED_[GLOBAL|LOCAL]_UNIT_ID

devreal committed:
Add DASH_UNDEFINED_[GLOBAL|LOCAL]_UNIT_ID

devreal committed:
Use typed unit IDs in BlockPattern*

devreal committed:
Use typed unit IDs in BlockPattern*

devreal committed:
Convert remaining patterns to typed pattern IDs

devreal committed:
Convert remaining patterns to typed pattern IDs

devreal committed:
Adjust TeamSpec to use typed unit IDs

devreal committed:
Adjust TeamSpec to use typed unit IDs

devreal committed:
Fix type of DART_UNDEFINED_GLOBAL_UNIT_ID and rename to dash::UNDEFINED_GLOBAL_UNIT_ID

devreal committed:
Fix type of DART_UNDEFINED_GLOBAL_UNIT_ID and rename to dash::UNDEFINED_GLOBAL_UNIT_ID

devreal committed:
Adapt PatternConcept to typed unit IDs

devreal committed:
Adapt PatternConcept to typed unit IDs

devreal committed:
Adapt dash::Array to typed unit IDs and fix ctor

devreal committed:
Adapt dash::Array to typed unit IDs and fix ctor

devreal committed:
Do not store the local ID in the Matrix

devreal committed:
Do not store the local ID in the Matrix

devreal committed:
Adapt some more containers to typed IDs

devreal committed:
Adapt some more containers to typed IDs

devreal committed:
Adapt *Locality implementations to typed unit IDs

devreal committed:
Adapt *Locality implementations to typed unit IDs

devreal committed:
Convert most remaining usages of dart_unit_t to typed unit IDs

devreal committed:
Convert most remaining usages of dart_unit_t to typed unit IDs

devreal commented:
Please carefully review and consider a squash merging

fuchsto commented:
I'm on it, thanks @devreal

fuchsto committed:
Merged branch feat-143-unitids

fuchsto committed:
Merge branch 'feat-143-unitids' of github.com:dash-project/dash into feat-143-unitids

Conflicts:
	dash/test/HDF5ArrayTest.h
	dash/test/HDF5MatrixTest.h

fmoessbauer commented on dash/examples/ex.02.matrix/main.cpp:
```diff
@@ -40,7 +40,7 @@ int main(int argc, char* argv[])
   teamspec_2d.balance_extents();
 
-  dart_unit_t myid   = dash::myid();
+  dash::global_unit_t myid   = dash::myid();
```
Why is myid of `global_unit_t`? Does that even matter for this example?

fmoessbauer commented on dash/include/dash/GlobMem.h:
```diff
@@ -253,13 +253,13 @@ class GlobMem
    * \param global_unit_id id of unit in \c dash::Team::All()
    */
   const ElementType * lbegin(
```
A globmem is allocated with a team, hence its (at least) public methods should use local unit ids. @fuchsto , @devreal , @fuerlinger please correct me if I am wrong.

I know, this will create a cascade of changes in DASH.

devreal replied:
I agree, this is a point I noticed but forgot to mention in the discussion. We should fix this to use local_unit_t.

fuchsto replied:
@devreal @fmoessbauer @fuerlinger See my comment here:

https://github.com/dash-project/dash/issues/143#issuecomment-265541969

fuchsto committed:
Testing NastyMPI CI build on development branch

fmoessbauer commented on dash/include/dash/SharedCounter.h:
```diff
@@ -22,7 +22,7 @@ class SharedCounter {
    */
```
The `SharedCounter` should either be deleted or reimplemented, as it is not team aware. I will open an issue for that.

fuchsto committed:
Merge branch 'development' of github.com:dash-project/dash into feat-143-unitids

fuchsto committed:
Added test suite base class dash::test::TestBase

fuchsto committed:
Defined move, copy, assignment as deleted for dash::Array

fuchsto committed:
Disabled MPI shared windows in Minimal, Nasty CI builds

fuchsto committed:
Added unit test to document defective conversion of local/global unit ids

fuchsto committed:
Added unit test to document defective conversion of local/global unit ids

fuchsto commented:
The changes need documentation.

`dash::internal::unit_id` should not be located in internal namespace as it would be hidden from the public interface otherwise.

To users, the interface of unit id types is highly relevant, however. Use cases like:

~~~c++
for (dash::local_unit_t lid { 0 }; lid < nunits; ++lid) { ... }
~~~

... are not self-explanatory if the operator overloads are not documented.

More important, type conversion operators should be defined as `= delete`.
Currently there is a conversion operator to `dart_unit_t` and a conversion constructor from `dart_unit_t`.

In effect, this allows use cases we intended to prohibit in the first place:

~~~c++
dash::local_unit_t  l_uid { 12 };
dash::global_unit_t g_uid { 12 };
l_uid = g_uid; // works, but shouldn't even compile
~~~

Added test suite `UnitIdTest` for further test of unit id semantics.

devreal commented:
We can move the unit_id struct to the public interface, no problem. I will also add documentation for the type. 

The reason for adding the conversion operator to dart_unit_t is to avoid having to access .id when calling dart_* functions. The problem is not the conversion c'tor but the assignment operator. I will delete it. 

I am a bit unhappy that we now have to write `dash::local_unit_t lid { 0 };` instead of a simple assignment but allowing implicit construction from `dart_unit_t` would prevent us from catching cases of wrong usage if the raw type is passed to a DASH function.

fuchsto commented:
You can, but

~~~c++
operator dart_gptr_t() noexcept
~~~

... should be explicit then.

Otherwise an implicit conversion `local_unit_t -> dart_gptr_t -> global_unit_t` would be allowed.
Assignments like

~~~c++
dash::local_unit_t lid = 12;
~~~

... should also be okay then, but

~~~c++
dash::local_unit_t lid = dash::global_unit_t(12);
~~~

... should not, while

~~~c++
dash::local_unit_t lid = static_cast<dart_gptr_t>(dash::global_unit_t(12));
~~~

... is okay again.

devreal commented:
Mhh, making 
```
operator dart_unit_t() noexcept
```
explicit seems to also prevent simple comparison with integer operands, e.g., `_myid == -1`. That would require an even bigger change to the code base.

fuchsto commented:
Number literals in code have implicit type `int`, we can rely on that.
So specializing operators for `int` should help.

Ah, but then again, `dart_gptr_t` *is* an `int`.
I'll find a solution, I adore stuff like that.

devreal commented:
Unfortunately, `dart_unit_t` is a `typedef` to `int32_t` (which in fact is `int`). The compiler does not honor different typedef'd types. So that won't help either.

fuchsto commented:
Yes, noticed that a wee moment later, see my edited comment.

I will add some unit tests for discussion.

devreal commented:
It seems that we can actually delete 
```
template<unit_scope Scope>
unit_id<IdScope> operator=(const unit_id<Scope> & id)
```

Have to test that though.

devreal committed:
Prohibit assignement from other unit_id types than the current type

devreal committed:
Prohibit assignement from other unit_id types than the current type

fuchsto committed:
Fixed deallocation order in dart_exit

devreal committed:
Move unit_id to dash:: namespace

devreal committed:
Move unit_id to dash:: namespace

fuchsto committed:
Stability fixes in global allocation, dash::Array

fuchsto committed:
Enabled shared MPI windows in CI build (bug in OpenMPI 2)

fuchsto committed:
CI: disabled OpenMPI-2 target, using correct MPI shared window settings

fmoessbauer commented on dash/include/dash/Team.h:
```diff
@@ -542,7 +543,7 @@ class Team
   size_t                  _num_siblings = 0;
   mutable size_t          _size         = 0;
-  mutable dart_unit_t     _myid         = -1;
+  mutable local_unit_t    _myid{-1};
```
Use better `UNDEFINED_LOCAL_UNIT_ID`

fmoessbauer commented on dash/include/dash/algorithm/Find.h:
```diff
@@ -47,7 +47,7 @@ GlobIter<ElementType, PatternType> find(
     // Pointer to first element in local memory:
     const ElementType * lbegin        = first.globmem().lbegin(
-                                           team.myid());
+                                           dash::myid());
```
Use `dash::Team::GlobalUnitID()`

fmoessbauer commented on dash/include/dash/algorithm/LocalRange.h:
```diff
@@ -269,7 +269,7 @@ local_range(
   auto pattern = first.pattern();
   auto lbegin  = first.globmem().lbegin(
-                   pattern.team().myid());
+                   dash::myid());
```
Use `dash::Team::GlobalUnitID()`

fmoessbauer commented on dash/include/dash/algorithm/MinMax.h:
```diff
@@ -191,7 +191,7 @@ GlobIter<ElementType, PatternType> min_element(
     // Pointer to first element in local memory:
     const ElementType * lbegin        = first.globmem().lbegin(
-                                          team.myid());
+                                          dash::myid());
```
Use `dash::Team::GlobalUnitID()`

fmoessbauer commented on dash/include/dash/iterator/GlobIter.h:
```diff
@@ -304,7 +304,7 @@ class GlobIter
     // Global reference to element at given position:
     return ReferenceType(
-             _globmem->at(local_pos.unit,
+             _globmem->at(dash::local_unit_t(local_pos.unit),
```
`dash` not necessary, so remove it

devreal replied:
Good catch!

fmoessbauer commented on dash/include/dash/util/UnitLocality.h:
```diff
@@ -63,12 +63,12 @@ class UnitLocality
-  : UnitLocality(dash::Team::All(), unit)
+    global_unit_t   unit)
+  : UnitLocality(dash::Team::All(), local_unit_t(unit))
   { }
```
Seems strange to me. Why `Team::All` and local id's ?

devreal replied:
The idea here is to enforce the use of a global ID if the team is guaranteed to be `dash::Team::All()`.

devreal replied:
Sorry, got your questions wrong: dash::Team::All() is a team like any other, so we have to pass a team-local unit ID to this constructor. We cannot handle the difference between `dash::Team::All()` and arbitrary teams with the types we have (unfortunately).

fmoessbauer commented on dash/test/GlobMemTest.cc:
```diff
@@ -6,7 +6,7 @@ TEST_F(GlobMemTest, ConstructorInitializerList)
 
   std::vector<int> glob_values;
-  for (int u = 0; u < dash::size(); u++) {
+  for (dash::local_unit_t u{0}; u < dash::size(); u++) {
```
Use `dash::global_unit_t` as team splits would break that loop.

devreal replied:
How would a team-split happen in this loop?

fmoessbauer replied:
Sorry, thats my mistake. I mixed up types and semantics. I thought that a local unit type has a limited range (but as it is constructed without a team that is of course not right).
However for me it still looks strange as `dash::size()` refers to a "global" size, but currently we do not have something like local and global sizes. I do not even think that this might be beneficial at any place...

Simply ignore my comment. Your code is correct 👍

fmoessbauer reviewed (commented):
Hi, nice job!
Just some minor points mentioned in the comments.

devreal committed:
Comment out statement that now fails to compile in UnitIdTest

devreal committed:
Comment out statement that now fails to compile in UnitIdTest

devreal committed:
Add documentation for unit_id and its specializations; Delete mixed-type comparison operators

devreal committed:
Add documentation for unit_id and its specializations; Delete mixed-type comparison operators

devreal commented:
Thanks a lot for your comments. I will fix the issues tomorrow.

fuchsto committed:
Applied fix from d98f63a, tentative rollback of dash::GlobMem

fuchsto committed:
Stability improvements for GlobMem tests

Commit:
Merge pull request #189 from dash-project/feat-52-dev-nastympi

Applying integration of NastyMPI CI build

fuchsto committed:
Removed Open Hub badge as it reports false data

fuchsto committed:
Specifying the MPI implementation in README

devreal committed:
Make creation of handles more robust

devreal committed:
Introduce DART_TEAM_NULL

devreal committed:
Introduce DART_TEAM_NULL

devreal committed:
Remove explicit cast that is not needed anymore (spotted by fmoesbauer)

devreal committed:
Remove explicit cast that is not needed anymore (spotted by fmoesbauer)

devreal committed:
Internally, use dash::Team::GlobalUnitID() instead of dash::myid()

devreal committed:
Internally, use dash::Team::GlobalUnitID() instead of dash::myid()

fuchsto committed:
Merge pull request #190 from dash-project/bug-dart-handles

Make handling of handles more robust

fmoessbauer committed:
make conversion operator to dart_unit_t const to enhance c++14 compability

fmoessbauer committed:
make conversion operator to dart_unit_t const to enhance c++14 compability

fmoessbauer committed:
Use dash::myid() in logger, as forward decls of static Team::GlobalUnitId() are not possible

fmoessbauer committed:
Use dash::myid() in logger, as forward decls of static Team::GlobalUnitId() are not possible

fmoessbauer committed:
Feat ex stencil (#196)

implemented ex.11.simple-stencil which simulates a blur filter applied to an image

fuchsto committed:
Added dockerhub badges.

fuchsto committed:
Changed labels of docker badges

BenjaProg committed:
After debugging issue #203 ddiefenthaler found out that
there is no dart_datatype for unsigned char.
So it would be a new feature perhaps.

With the changes in this commit the initial example
in dash-apps will run as intended with the datatype unsigned char as
well.
But we can not say if the changes suffice for the general case.

Please review and remark!

fmoessbauer committed:
changed order of types to be consisten with the other types

fmoessbauer committed:
enhanced documentation of dash::transform (#205)

devreal committed:
Several fixes related to typed unit IDs; Implement dash::Team::relative_id(global_unit_t) and use it instead of related DART functions

devreal committed:
Several fixes related to typed unit IDs; Implement dash::Team::relative_id(global_unit_t) and use it instead of related DART functions

devreal committed:
Implement typed unit IDs in DART and make necessary adjustments in DASH; Please carefully review and fix TODO[TF]

devreal committed:
Implement typed unit IDs in DART and make necessary adjustments in DASH; Please carefully review and fix TODO[TF]

devreal committed:
Merged upstream

devreal committed:
Fixes for Clang build

devreal committed:
Fixes for Clang build

devreal committed:
Fix invocation of template-function sub<N> in DASH Matrix

devreal committed:
Fix reference to static *Pattern::PatterName

devreal committed:
Fix potential use of uninitialized variable reported by Clang

devreal committed:
Fix faulty use of static function Pattern::ndim() reported by Clang

devreal committed:
Fix GlobRef::operator!=()

devreal committed:
Remove unused variable

devreal committed:
Remove unnecessary temporary object to calm Clang warning

devreal committed:
Only include omp.h if necessary

devreal committed:
Fix pedantic warning on format type mismatch

devreal committed:
Fix wrong use of static dash::Matrix::ndim() in HDF5 StorageDriver

devreal committed:
Some more fixes for use of static PatternName member

devreal committed:
Fix header guard

devreal committed:
Remove _parent pointer from LocalityDomain
reported unused by Clang

devreal committed:
Use static_cast instead of C-style cast

devreal committed:
Add documentation for unit ID types

devreal committed:
Add documentation for unit ID types

devreal committed:
Merge pull request #207 from dash-project/bug-206-clang

Fix errors and warnings reported by Clang

devreal committed:
Merge pull request #204 from dash-project/feat-203-DartTypeUChar

Add DART data type for unsigned char.

devreal committed:
Use a copy of MPI_COMM_WORLD to avoid conflicts on global collective operations (#210)

* Use a copy of MPI_COMM_WORLD (DART_COMM_WORLD) to avoid conflicts on global collective operations

* Make dart__base__locality__* fields static

* Fix move assignemnt operators/ctors and avoid direct use of DART_TEAM_NULL

* Reintroduce deleted nullptr init

fuchsto committed:
Merge branch 'feat-143-unitids' of github.com:dash-project/dash into feat-143-unitids

Conflicts:
	dart-impl/mpi/src/dart_communication.c
	dart-impl/mpi/src/dart_initialization.c
	dash/include/dash/internal/Unit.h

fuchsto committed:
Using local unit ids in dart_domain_locality_t

fuchsto committed:
Added missing cast constructor in dash::unit_id

fuchsto committed:
Applied interface changes of DART/DASH locality modules

fuchsto committed:
Using global unit ids in dart_domain_locality_t

fuchsto committed:
Fixed local/global unit id conversions in DART locality module

fuchsto committed:
Removed openmpi2 build from pulled containers in CI

fuchsto committed:
Fixed log message in test case

fuchsto committed:
Resolved review requests, renamed *local_unit_t to *team_unit_t

fuchsto committed:
Renamed dependent definitions of *local_unit_t to *team_unit_t

fuchsto commented:
@devreal
We should be ready for merging this behemoth.

Summary of my changes:

- Rebased to recent state of development branch
- Updated implementation of DART/DASH locality modules to typed unit ids
- Resolved your requests for review (marked `TODO[TF]` in comments)
- Renamed `*local_unit_t` to `*team_unit_t` and dependent definitions like `DART_UNDEFINED_LOCAL_UNIT_ID` / `DART_UNDEFINED_TEAM_UNIT_ID` etc.

**Please use a regular merge** (as in: **not** squash & merge) as the commit history is very, very valuable for future bug hunts.

devreal committed:
Fix some TODOs and make Clang happy
